crates/core_simd/src/vector.rs

```diff
@@ -979,3 +1134,27 @@ where
 {
     type Mask = isize;
 }
+
+#[inline]
+fn lane_indices<T, const N: usize>() -> Simd<T, N>
+where
+    T: MaskElement + Default + core::convert::From<i8> + core::ops::Add<T, Output = T>,
+    LaneCount<N>: SupportedLaneCount,
+{
+    let mut index = [T::default(); N];
+    for i in 1..N {
+        index[i] = index[i - 1] + T::from(1);
+    }
+    Simd::from_array(index)
+}
+
+#[inline]
+fn mask_up_to<M, const N: usize>(len: usize) -> Mask<M, N>
+where
+    LaneCount<N>: SupportedLaneCount,
+    M: MaskElement,
+{
+    let index = lane_indices::<i8, N>();
+    let lt = index.simd_lt(Simd::splat(i8::try_from(len).unwrap_or(i8::MAX)));
+    Mask::<M, N>::from_bitmask_vector(lt.to_bitmask_vector())
```

Comment:
Maybe it has been considered. any idea to avoid calling `lane_indices` (and no need to run the for loop inside the function each time the `mask_up_to` is called), by :

```rust
fn mask_up_to<...>() -> Mask<M,N> ... {
    Mask::<M, N>::from_bitmask((1_u64 << len) - 1)
}
```

pls let me know if I misunderstood the function :)

Reply:
The previous version of this PR had the option to use this exact bitmask and indeed, it tends to produce better performing code https://github.com/rust-lang/portable-simd/pull/374#issuecomment-1817863838

Plus, it's just simpler, so I'm in favor. Any objections @calebzulawski? This would not require adding any extra trait bounds ever since you removed ToBitMask.

Reply:
Oh, and it would have to be this to prevent an overflow on the shift:

```rust
fn mask_up_to<...>() -> Mask<M,N> ... {
    if len >= 64 {
        Mask::<M, N>::splat(true)
    } else {
        Mask::<M, N>::from_bitmask((1_u64 << len) - 1)
    }
}
```

Reply:
it might be...if desirable to avoid the literal "64"
```rust
 if len >= u64::BITS as usize { ...
```

Reply:
I think you need to condition that on x86, because that will be worse on other architectures.  Also, you should condition that on the lane count being no more than 64, because we will support larger in the future. Alternatively, use a bitmask vector which works for larger than 64.

Reply:
any comment if giving shortcuts when `len` not less than lane count?

```rust
fn mask_up_to<...>() -> Mask<M,N> ... {
    if len >= N {
        Mask::<M,N>::splat(true)
    }else{
        let index = lane_indices::<i8, N>();
        let lt = index.simd_lt(Simd::splat(i8::try_from(len).unwrap_or(i8::MAX)));
        // .... Mask::<M, N>::from_bitmask_vector(lt.to_bitmask_vector())
    }
}
```

Reply:
I think the comparison based version also generates better code on non-avx512 machines: https://rust.godbolt.org/z/6edK1rx9M

Regarding the initial question, since the length parameter is a const generic we would expect llvm to always precompute the mask vector, so at runtime it is a single vector load.

Reply:
We're at risk of having the same conversation as in the previous PR, which we wanted to avoid with this spin-off PR.

Let's go with the bitmask vector implementation, which is more generic. We can revisit this later, or LLVM can learn to optimize this better, or users who are performance-conscious today can use the unsafe functions to provide their own mask, calculated in a way that best fits their architecture and usecase.